src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +167,268 @@ func (b *dbBuffer) Write(
-				)
-			}
-			b.buckets[i].bootstrap(bl)
-			bootstrapped = true
-			break
+	for iter.Next() {
+		dp, unit, annotation := iter.Current()
+		if err := encoder.Encode(dp, unit, annotation); err != nil {
+			return nil, err
 		}
 	}
-	if !bootstrapped {
-		return fmt.Errorf("block at %s not contained by buffer", blockStart.String())
+	if err := iter.Err(); err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-// forEachBucketAsc iterates over the buckets in time ascending order
-// to read bucket data
-func (b *dbBuffer) forEachBucketAsc(fn func(*dbBufferBucket)) {
-	for i := 0; i < bucketsLen; i++ {
-		idx := (b.pastMostBucketIdx + i) % bucketsLen
-		fn(&b.buckets[idx])
-	}
+	return encoder.Stream(), nil
 }
 
-// computedForEachBucketAsc performs a fn on the buckets in time ascending order
-// and returns the sum of the number returned by each fn
-func (b *dbBuffer) computedForEachBucketAsc(
-	op computeBucketIdxOp,
-	fn func(now time.Time, b *dbBuffer, idx int, bucketStart time.Time) int,
-) int {
-	now := b.nowFn()
-	pastMostBucketStart := now.Truncate(b.blockSize).Add(-1 * b.blockSize)
-	bucketNum := (pastMostBucketStart.UnixNano() / int64(b.blockSize)) % bucketsLen
-	result := 0
-	for i := int64(0); i < bucketsLen; i++ {
-		idx := int((bucketNum + i) % bucketsLen)
-		curr := pastMostBucketStart.Add(time.Duration(i) * b.blockSize)
-		result += fn(now, b, idx, curr)
-	}
-	if op == computeAndResetBucketIdx {
-		b.pastMostBucketIdx = int(bucketNum)
+func (b *dbBuffer) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	id ident.ID,
+	tags ident.Tags,
+	persistFn persist.DataFn,
+	version int,
+) (FlushOutcome, error) {
+	buckets, exists := b.bucketVersionsAt(blockStart)
+	if !exists {
+		return FlushOutcomeBlockDoesNotExist, nil
 	}
-	return result
-}
 
-func (b *dbBuffer) Snapshot(ctx context.Context, blockStart time.Time) (xio.SegmentReader, error) {
-	var (
-		res xio.SegmentReader
-		err error
-	)
+	streams, err := buckets.toStreams(ctx)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
+	var stream xio.SegmentReader
+	if numStreams := len(streams); numStreams == 1 {
+		stream = streams[0]
+	} else {
+		// In the majority of cases, there will only be one stream to persist
+		// here. Only when a previous flush fails midway through a shard will
+		// there be buckets for previous versions. In this case, we need to try
+		// to flush them again, so we merge them together to one stream and
+		// persist it.
+		encoder, _, err := mergeStreamsToEncoder(blockStart, streams, b.opts)
 		if err != nil {
-			// Something already went wrong and we want to return the error to the caller
-			// as soon as possible instead of continuing to do work.
-			return
+			return FlushOutcomeErr, err
 		}
 
-		if !bucket.canRead() {
-			return
-		}
+		stream = encoder.Stream()
+		encoder.Close()
+	}
 
-		if !blockStart.Equal(bucket.start) {
-			return
-		}
+	segment, err := stream.Segment()
```

Comment:
Does this segment ever get finalized?  Seems like it should get finalized before this method finishes but it doesn't seem to?

Reply:
Hmm yeah that makes sense. I think I was modeling based off of the `Flush` func in series [here](https://github.com/m3db/m3/blob/master/src/dbnode/storage/series/series.go#L617). That doesn't seem to finalize the segment either - is that different somehow or should that finalize the segment as well?

Reply:
So I added some comments to the buffer `toStream(...)` method.  If you make those changes then we don't need to finalizer here (which is better than actually finalizing here).